File: src/main/java/com/manson/domain/fo76/items/enums/FilterFlag.java
package com.manson.domain.fo76.items.enums;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import org.apache.commons.lang3.StringUtils;

public enum FilterFlag {
    POWER_ARMOR("Power Armor", 0L),
    WEAPON_MELEE("Weapon - Melee", true, false),
    WEAPON_RANGED("Weapon - Ranged", true, false),
    WEAPON_THROWN("Weapon - Thrown", false, 2L, 3L),
    WEAPON("Weapon", true, Arrays.asList(WEAPON_THROWN, WEAPON_MELEE, WEAPON_RANGED), RawFilterFlags.FILTER_WEAPONS,
        5L),
    ARMOR("Armor", true, Collections.emptyList(), RawFilterFlags.FILTER_ARMOR),
    APPAREL("Apparel", RawFilterFlags.FILTER_APPAREL),
    FOOD_WATER("Food - Water", RawFilterFlags.FILTER_FOODWATER, 33L, 9L),
    AID("Aid", RawFilterFlags.FILTER_AID, 65L),
    PLANS("Plans", false, false),
    NOTES("Notes", false, Collections.singletonList(PLANS), RawFilterFlags.FILTER_BOOKS, 128L),
    HOLO("Holo", RawFilterFlags.FILTER_HOLOTAPES),
    AMMO("Ammo", RawFilterFlags.FILTER_AMMO),
    MISC("Misc", RawFilterFlags.FILTER_MISC, 266240L, 512L, 33280L),
    MODS("Mods", RawFilterFlags.FILTER_MODS, 2048L),
    JUNK("Junk", RawFilterFlags.FILTER_JUNK, 270336L, 33792L),
    UNKNOWN("Unknown");

    private final List<Long> flags = new ArrayList<>();
    private final String value;
    private final boolean hasStarMods;
    private final List<FilterFlag> subtypes;
    private final boolean parent;

    FilterFlag(String name, boolean hasStarMods, List<FilterFlag> subtypes, boolean parent, Long... flags) {
        this.flags.addAll(Arrays.asList(flags));
        this.value = name;
        this.hasStarMods = hasStarMods;
        this.subtypes = subtypes;
        this.parent = parent;
    }

    FilterFlag(String name, boolean hasStarMods, List<FilterFlag> subtypes, Long... flags) {
        this(name, hasStarMods, subtypes, true, flags);
    }

    FilterFlag(String name, boolean parent, Long... flags) {
        this(name, false, Collections.emptyList(), parent, flags);
    }

    FilterFlag(String name, Long... flags) {
        this(name, false, Collections.emptyList(), true, flags);
    }

    FilterFlag(String name, boolean hasStarMods, boolean parent) {
        this(name, hasStarMods, Collections.emptyList(), parent);
    }

    FilterFlag(String name, boolean hasStarMods) {
        this(name, hasStarMods, Collections.emptyList(), true);
    }

    public static FilterFlag fromInt(Long flag) {
        if (flag == null) {
            return UNKNOWN;
        }
        return Arrays.stream(values())
            .filter(filterFlag -> filterFlag.getFlags().contains(flag)).findFirst().orElse(UNKNOWN);
    }

    public static FilterFlag fromString(String flag) {
        if (StringUtils.isBlank(flag)) {
            return null;
        }
        return Arrays.stream(values())
            .filter(filterFlag -> StringUtils.equalsIgnoreCase(flag, filterFlag.value) || StringUtils
                .equalsIgnoreCase(flag, filterFlag.name())).findFirst().orElse(null);
    }

    public List<Long> getFlags() {
        return flags;
    }

    public String getValue() {
        return value;
    }

    public boolean isHasStarMods() {
        return hasStarMods;
    }

    public List<FilterFlag> getSubtypes() {
        return subtypes;
    }

    public boolean isParent() {
        return parent;
    }
}

File: src/main/kotlin/com/manson/domain/fo76/items/enums/FilterFlag.kt
package com.manson.domain.fo76.items.enums

import java.util.ArrayList
import java.util.Arrays
import org.apache.commons.lang3.StringUtils

enum class FilterFlag {
    POWER_ARMOR("Power Armor", 0),
    WEAPON_MELEE("Weapon - Melee", true),
    WEAPON_RANGED("Weapon - Ranged", true),
    WEAPON_THROWN("Weapon - Thrown"),
    WEAPON("Weapon", true, listOf(WEAPON_THROWN, WEAPON_MELEE, WEAPON_RANGED), 2, 3),
    ARMOR_OUTFIT("Armor - Outfit"),
    ARMOR("Armor", true, listOf(ARMOR_OUTFIT), 4),
    AID("Aid", 8, 9),
    HOLO("Holo", 512),
    AMMO("Ammo", 4096),
    NOTES("Notes", 128, 8192),
    MISC("Misc", 33280),
    MODS("Mods", 2048),
    JUNK("Junk", 33792, 1024),
    UNKNOWN("Unknown");

    private val flags: MutableList<Int> = ArrayList()
    val value: String
    val isHasStarMods: Boolean
    val subtypes: List<FilterFlag>

    constructor(name: String, hasStarMods: Boolean, subtypes: List<FilterFlag> = listOf(), vararg flags: Int) {
        this.flags.addAll(flags.asList())
        this.value = name
        this.isHasStarMods = hasStarMods
        this.subtypes = subtypes
    }

    constructor(name: String, vararg flags: Int) : this(name, hasStarMods = false, flags = flags)

    fun getFlags(): List<Int> {
        return flags.toList()
    }

    companion object {
        @JvmStatic
        fun fromInt(flag: Int?): FilterFlag {
            return if (flag == null) {
                UNKNOWN
            } else Arrays.stream(values())
                    .filter { filterFlag: FilterFlag -> filterFlag.getFlags().contains(flag) }.findFirst().orElse(
                    UNKNOWN
                )
        }

        @JvmStatic
        fun fromString(flag: String?): FilterFlag {
            return if (StringUtils.isBlank(flag)) {
                UNKNOWN
            } else Arrays.stream(values())
                    .filter { filterFlag: FilterFlag -> StringUtils.equalsIgnoreCase(flag, filterFlag.value) }.findFirst().orElse(
                    UNKNOWN
                )
        }
    }
}